crypto/x509/t_acert.c

```diff
@@ -0,0 +1,292 @@
+        type = at->type;
+
+        switch (type) {
+        case V_ASN1_PRINTABLESTRING:
+        case V_ASN1_T61STRING:
+        case V_ASN1_NUMERICSTRING:
+        case V_ASN1_UTF8STRING:
+        case V_ASN1_IA5STRING:
+            bs = at->value.asn1_string;
+            if (BIO_write(bp, (char *)bs->data, bs->length) != bs->length)
+                goto err;
+            if (BIO_puts(bp, "\n") <= 0)
+                goto err;
+            break;
+        case V_ASN1_SEQUENCE:
+            if (BIO_puts(bp, "\n") <= 0)
+                goto err;
+            ASN1_parse_dump(bp, at->value.sequence->data,
+                            at->value.sequence->length, i, 1);
+            break;
+        default:
+            if (BIO_printf(bp, "unable to print attribute of type 0x%X\n",
+                           type) < 0)
+                goto err;
+            break;
+        }
+    }
+    ret = 1;
+err:
+    return ret;
+}
+
+int X509_ACERT_print_ex(BIO *bp, X509_ACERT *x, unsigned long nmflags,
+                        unsigned long cflag)
+{
+    int i;
+    char mlch = ' ';
+
+    if ((nmflags & XN_FLAG_SEP_MASK) == XN_FLAG_SEP_MULTILINE) {
+        mlch = '\n';
+    }
+
+    if ((cflag & X509_FLAG_NO_HEADER) == 0) {
+        if (BIO_printf(bp, "Attribute Certificate:\n") <= 0)
+            goto err;
+        if (BIO_printf(bp, "%4sData:\n", "") <= 0)
+            goto err;
+    }
+
+    if ((cflag & X509_FLAG_NO_VERSION) == 0) {
+        long l;
+
+        l = X509_ACERT_get_version(x);
+        if (l == X509_ACERT_VERSION_2) {
+            if (BIO_printf(bp, "%8sVersion: %ld (0x%lx)\n", "", l + 1,
+                           (unsigned long)l) <= 0)
+                goto err;
+        } else {
+            if (BIO_printf(bp, "%8sVersion: Unknown (%ld)\n", "", l) <= 0)
+                goto err;
+        }
+    }
+
+    if ((cflag & X509_FLAG_NO_SERIAL) == 0) {
+        const ASN1_INTEGER *serial;
+        char *serial_str;
+
+        serial = X509_ACERT_get0_serialNumber(x);
+        serial_str = i2s_ASN1_INTEGER(NULL, serial);
+        if (serial_str == NULL)
+            goto err;
+
+        if (BIO_printf(bp, "%8sSerial Number: %s\n", "", serial_str) <= 0) {
+            OPENSSL_free(serial_str);
+            goto err;
+        }
+        OPENSSL_free(serial_str);
+    }
+
+    if ((cflag & X509_FLAG_NO_SUBJECT) == 0) {
+        const GENERAL_NAMES *holderEntities;
+        const OSSL_ISSUER_SERIAL *holder_bcid;
+        const X509_NAME *holderIssuer = NULL;
+
+        if (BIO_printf(bp, "%8sHolder:\n", "") <= 0)
+            goto err;
+
+        holderEntities = X509_ACERT_get0_holder_entityName(x);
+        if (holderEntities != NULL) {
+            GENERAL_NAME *entity = NULL;
+            /*
+             * Just display the first name in the sequence.  Any others
+             * are ignored.
+             */
+            if (sk_GENERAL_NAME_num(holderEntities) >= 1)
+                entity = sk_GENERAL_NAME_value(holderEntities, 0);
```

Comment:
I think it is problematic to have a pretty-printer which doesn't show all of the information and just silently hides it.

Reply:
Come to think of it, I also wonder what would happen if this received an empty SEQUENCE. Segfault?